### PR TITLE
apiserver: graceful termination with new option shutdown-send-retry-after

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Introduce a new server run option `shutdown-send-retry-after` 
```
If true the HTTP Server will continue listening until all non longrunning request(s) in flight 
have been drained, during this window all incoming requests will be rejected with
a status code 429 and a 'Retry-After' response header.
```
The default value is set to `false` to maintain current behavior.

With this option, we now have two modes of graceful termination:
- `shutdown-send-retry-after=true`: we initiate shutdown of the HTTP Server  when all non `longrunning` requests in-flight have been drained.
- `shutdown-send-retry-after=false`: we initiate shutdown of the HTTP Server  as soon as `shutdown-delay-duration` has elapsed. (This is the current behavior)

The following happens with `shutdown-send-retry-after=true`:
- Do not call http `Server.Shutdown` immediately after  `ShutdownDelayDuration` elapses. Rather, wait for all non `longrunning` requests to drain. (the requests that are being counted by `WithWaitGroup` filter)
- Immediately after `ShutdownDelayDuration` elapses, any new incoming request  should be answered with a `429` and the following headers: 
- `Connection: close`  and 
- `Retry-After: N`

#### Does this PR introduce a user-facing change?
```release-note
a new server run option 'shutdown-send-retry-after'  has been introduced. If true the HTTP Server
will continue listening until all non longrunning request(s) in flight have been drained, during this window all 
incoming requests will be rejected with a status code 429 and a 'Retry-After' response header.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by smarterclayton** [comment]
Hrm, I'd prefer 429 - try again (every request after shutdown we reject is " come back later) and connection close ensures that.

**Comment by tkashem** [comment]
> Hrm, I'd prefer 429 - try again (every request after shutdown we reject is " come back later) and connection close ensures that.

@smarterclayton  (this was a typo in my PR description) yes, we return `429`

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 1]
```diff
@@ -233,6 +233,18 @@ type Config struct {
...
 	// StorageVersionManager holds the storage versions of the API resources installed by this server.
 	StorageVersionManager storageversion.Manager
+
+	// DelayServerShutdown dictates when to initiate shutdown of the HTTP Server
+	// during the graceful termination of the apiserver.
+	// If true, we wait for existing requests in flight to be completed and
+	// then initiate a shutdown of the HTTP Server.
+	// If false, we initiate a shutdown of the HTTP Server
+	// as soon as ShutdownDelayDuration has elapsed.
+	DelayServerShutdown bool
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:243]: Can we invert the logic in the var name and the flag? `keepListeningDuringGracefulTermination`.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -233,6 +233,18 @@ type Config struct {
...
+
+	// GracefulShutdown provides access to the various shutdown events
+	// that happen during the graceful termination of the apiserver.
+	GracefulShutdown GracefulShutdown
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:247]: is this nil before shutdown?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:247]: wondering that this thing is and how to name it better. It is not "a shutdown".

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+// if the specified notAcceptingNewRequest channel is closed
+//
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving after notAcceptingNewRequest is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: 1` (so client can retry, hopefully on a new apiserver instance)
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithNotAcceptingNewRequest(handler http.Handler, notAcceptingNewRequest <-chan struct{}) http.Handler {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go:33]: <bike-shedding>With429After</bike-shedding> ?

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		var reject bool
+		select {
+		case <-notAcceptingNewRequest:
+			reject = true
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go:38]:
```
handler.ServeHTTP(w, req)
return
```

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 2 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:247]: it is not `nil`, we create a new instance of it inside `config.Complete`. we need it instanced up front so we can use the events in handler chain build function.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/config.go:247]: > wondering what this thing is and how to name it better. It is not "a shutdown".

`GracefulShutdownCoordinator`, `GracefulShutdownDriver`? let me know if you have any suggestions :)

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go (hunk 3 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_not_accepting.go:33]: How about `WithRetryAfter`?

**Comment by p0lyn0mial** [comment]
I am not sure if I understand why we need an additional option.

IIRC the maximum time a request can take is `60` seconds (`RequestTimeout`) (excluding long-running requests). So a request accepted at `T+70` will have been terminated at `T+70+60`.

What we could do is to:

- set the `ShutdownTimeout` to `70` seconds to give more time to the server to send a response for very late requests `(70+60)`

- send `429` immediately after starting the termination process to reduce the potential number of in-flight requests to handle.

I haven't seen code that actually determines the number of requests in-flight but it looks like a broken LB or a client that uses an API directly can significantly delay or even block termination of an API server. That seems very powerful.

File: staging/src/k8s.io/apiserver/pkg/server/graceful_shutdown.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+	// The GenericAPIServer should call Start once and then use the
+	// ShutdownEvents returned to coordinate a graceful termination.
+	//
+	// Start should be an idempotent call, calling it multiple times
+	// should not have any adverse effects.
+	Start(stopCh <-chan struct{}) ShutdownEvents
+
+	// Events returns the ShutdownEvents associated. The caller can
+	// can use to get notified when the shutdown events take place.
+	Events() ShutdownEvents
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/graceful_shutdown.go:162]: do we really need Events() and the return value of Start for the same thing? Looks like these two overlap.

**Comment by tkashem** [comment]
@sttts I am going to break it down to multiple PRs so it is easy to review - the refactor should be a separate PR and then a follow up PR for `keep-listening-during-graceful-termination`

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 2 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: addressed in the refactor pr

**Comment by tkashem** [comment]
no flakes so far:
```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress ./server.test -test.run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationEnabled
...
23m0s: 7240 runs so far, 0 failures
23m5s: 7268 runs so far, 0 failures
23m10s: 7296 runs so far, 0 failures
23m15s: 7320 runs so far, 0 failures
23m20s: 7348 runs so far, 0 failures
```

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 6]
```diff
@@ -351,7 +359,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
...
+	drainedCh := s.terminationSignals.InFlightRequestsDrained
+	stopHttpServerCh := delayedStopCh.Signaled()
+	shutdownTimeout := s.ShutdownTimeout
+	if s.KeepListeningDuringGracefulTermination {
+		// when this mode is enabled, we do the following:
+		// - the server will continue to listen until all existing
+		//   requests in flight have been drained
+		// - once drained, http Server Shutdown is invoked with a timeout of 5s
+		stopHttpServerCh = drainedCh.Signaled()
+		shutdownTimeout = 5 * time.Second
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:371]: it changes the previous behavior, it is not backward compatible, not sure it has to be
- it changes the maximum time from `60s` to `65s`
- doesn't apply the timeout set in `ShutdownTimeout` (could be set to 10s)

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 7]
```diff
@@ -785,6 +795,9 @@ func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
 	handler = genericapifilters.WithWarningRecorder(handler)
 	handler = genericapifilters.WithCacheControl(handler)
 	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
+	if c.KeepListeningDuringGracefulTermination {
+		handler = genericfilters.WithRetryAfter(handler, c.terminationSignals.AfterShutdownDelayDuration.Signaled())
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:799]: please add a comment that this handler must be **before** the `WithWaitGroup` otherwise the server can deadlock

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+// if the specified shutdownDelayDurationElapsedCh channel is closed
+//
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving after shutdownDelayDurationElapsedCh is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance)
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:33]: This filter changes the status codes previously returned on paths like `livez` and `readyz` previously `500` now it will be `429`. Again this might be not a backward compatible change/behavriour.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:33]: We might actually allow some actors to get through like `the loopback client`. Otherwise, the server might not shut down gracefully.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:33]: readyz/livez/healthyz/metrics definitely should not change

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:33]: +1 for loopback client

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:33]: I updated the filter to check the path and the user, PTAL https://github.com/kubernetes/kubernetes/pull/103357

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 9]
```diff
@@ -352,7 +360,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
...
 
 	// close socket after delayed stopCh
-	stoppedCh, listenerStoppedCh, err := s.NonBlockingRun(delayedStopCh.Signaled())
+	drainedCh := s.lifecycleSignals.InFlightRequestsDrained
+	stopHttpServerCh := delayedStopCh.Signaled()
+	shutdownTimeout := s.ShutdownTimeout
+	if s.KeepListeningDuringGracefulTermination {
+		// when this mode is enabled, we do the following:
+		// - the server will continue to listen until all existing
+		//   requests in flight have been drained
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:369]: be clear what this means for long-running requests

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 10]
```diff
@@ -352,7 +360,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
...
+		// - once drained, http Server Shutdown is invoked with a timeout of 5s
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:370]: what do the 5s mean? Why not immediately or 1s?

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:370]: I set it to `2s`, net/http waits for `1s` after sending GO_AWAY to take into account round trip time for the peer, so we should respect that.

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 11]
```diff
@@ -614,6 +622,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 		StorageVersionManager: c.StorageVersionManager,
 
 		Version: c.Version,
+
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:625]: nit: we have many empty lines in here that do not make sense :)

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+	Message string
+}
+
+// retryAfterConditionFunc returns true if the requests should
+// be rejected once certain conditions are met.
+// The retryAfterParams returned contains instructions on how to
+// construct the Retry-After response.
+type retryAfterConditionFunc func() (*retryAfterParams, bool)
+
+func (f retryAfterConditionFunc) ShouldSendRetryAfterResponse() (*retryAfterParams, bool) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:56]: what does "after" mean here? Is this `RespondWithRetryAfter`?

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+// It includes new request(s) on a new or an existing TCP connection
+// Any new request(s) arriving after shutdownDelayDurationElapsedFn is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance)
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+	return withRetryAfter(handler, func(r *http.Request) bool {
+		if hasExcludedPathPrefix(r) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:72]: `return hasExcludedPathPrefix || isKubeApiserverUserAgent`

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+			return true
+		}
+		return false
+	}, newRetryAfterFunc(shutdownDelayDurationElapsedCh))
+}
+
+func withRetryAfter(handler http.Handler, excludeFn excludeFunc, retryAfterFn retryAfterConditionFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := retryAfterFn.ShouldSendRetryAfterResponse()
+		if !sendRetryAfter {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:85]: if !sendRetryAfter || excludeFn(req)

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 10 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:370]: please add this in the comment. 5s seems to be arbitrary without explanation. Also sync comment and code :)

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 15]
```diff
@@ -352,7 +360,20 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
...
+	stopHttpServerCh := delayedStopCh.Signaled()
+	shutdownTimeout := s.ShutdownTimeout
+	if s.KeepListeningDuringGracefulTermination {
+		// when this mode is enabled, we do the following:
+		// - the server will continue to listen until all existing requests in flight
+		//   (not including active long runnning requests) have been drained.
+		// - once drained, http Server Shutdown is invoked with a timeout of 5s
+		stopHttpServerCh = drainedCh.Signaled()
+		shutdownTimeout = 2 * time.Second
+		klog.V(1).InfoS("[graceful-termination] HTTP Server", "ShutdownTimeout", shutdownTimeout)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:373]: The `[graceful-termination]` prefix is used everywhere?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go [hunk 16]
```diff
@@ -74,18 +74,18 @@ func (w *wrappedTerminationSignal) Signal() {
 	}
 }
 
-func wrapTerminationSignals(t *testing.T, ts *lifecycleSignals, callback func(bool, string, lifecycleSignal)) {
-	newWrappedTerminationSignal := func(delegated lifecycleSignal) lifecycleSignal {
-		return &wrappedTerminationSignal{
+func wrapLifecycleSignals(t *testing.T, ts *lifecycleSignals, callback func(bool, string, lifecycleSignal)) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:77]: factor these renames out into their own commit

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// if we are here this means it's time to send Retry-After response to
+		// all requests except the ones that are excluded.
+		if excludeFn(req) {
+			handler.ServeHTTP(w, req)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:93]: I think that it will keep the `drained` queue busy and will block the server from shutting down, am I right?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:93]: we can exlucde them from the wait group too

File: staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go [hunk 18]
```diff
@@ -245,5 +255,9 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"will return success, but /readyz immediately returns failure. Graceful termination starts after this delay "+
 		"has elapsed. This can be used to allow load balancer to stop sending traffic to this server.")
 
+	fs.BoolVar(&s.KeepListeningDuringGracefulTermination, "keep-listening-during-shutdown", s.KeepListeningDuringGracefulTermination, ""+
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:258]: sync with @p0lyn0mial about the names. They should look symmetric.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 15 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:373]: yes, so we can search with it during debugging

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:373]: I mean the syntax. It looks special. `Using graceful termination http server shutdown timeout` would be more normal.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go (hunk 12 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go]: renamed it to `ShouldRespondWithRetryAfter`

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go (hunk 17 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:80]: yeah, for now using the same exclude function in `WithWaitGroup`, I have left a todo to merge these two filters in the future.

File: staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go [hunk 19]
```diff
@@ -41,21 +47,32 @@ func WithWaitGroup(handler http.Handler, longRunning apirequest.LongRunningReque
...
+		if longRunning(req, requestInfo) {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		if err := wg.Add(1); err != nil {
+			// shutdown delay duration has elapsed and SafeWaitGroup.Wait has been invoked
+			// we are going to allow execution of the requests that WithRetryAfter is
+			// excluding from sending a Retry-After response.
+			if excludeFn(req) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go:59]: why is this not like long-running? Can we combine both predicates?

i.e. a function called excludeFunc in withWaitGroup and then it is not really excluded.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+
+import (
+	"net/http"
+	"strings"
+)
+
+var (
+	// health probes and metrics scraping are never rejected, we will continue
+	// serving these requests after shutdown delay duration elapses.
+	excludedPathPrefixes = []string{
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:27]: excluded from what? This variable is public for the whole filters package. It needs a more specialized name.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+
+// retryAfterConditionFunc returns true if the requests should
+// be rejected once certain conditions are met.
+// The retryAfterParams returned contains instructions on how to
+// construct the Retry-After response.
+type retryAfterConditionFunc func() (*retryAfterParams, bool)
+
+func (f retryAfterConditionFunc) ShouldRespondWithRetryAfter() (*retryAfterParams, bool) {
+	return f()
+}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:58]: odd construct. Just use `retryAfterConditionFunc()` directly. No need for a method.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+// TODO: is there a way to merge WithWaitGroup and this filter?
+// NOTE: both WithRetryAfter and WithWaitGroup must use the same exact excludeFunc 'isRequestExcluded,
+// otherwise SafeWaitGroup might wait indefinitely and will prevent the server from shutting down gracefully.
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+	return withRetryAfter(handler, isRequestExcluded, newRetryAfterFunc(shutdownDelayDurationElapsedCh))
+}
+
+func withRetryAfter(handler http.Handler, excludeFn excludeFunc, retryAfterFn retryAfterConditionFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := retryAfterFn.ShouldRespondWithRetryAfter()
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:78]: retryAfterFn()

File: staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go (hunk 19 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/waitgroup.go:59]: we don't want to combine with `longRunning` - `excludeFn` is relevant when shutdown delay duration has elapsed. Only then `WithRetryAfter` starts responding with retry-after response. 
inside `withWaitGroup` we are not sending retry-after for the white-listed (allowed) requests and we are not counting them either.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 23]
```diff
@@ -352,7 +361,22 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
...
+	shutdownTimeout := s.ShutdownTimeout
+	if s.ShutdownSendRetryAfter {
+		// when this mode is enabled, we do the following:
+		// - the server will continue to listen until all existing requests in flight
+		//   (not including active long runnning requests) have been drained.
+		// - once drained, http Server Shutdown is invoked with a timeout of 2s,
+		//   net/http waits for 1s for the peer to respond to a GO_AWAY frame, so
+		//   we should wait for a minimum of 2s
+		stopHttpServerCh = drainedCh.Signaled()
+		shutdownTimeout = 2 * time.Second
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:375]: the long-running connections should have a chance to terminate gracefully.
with `shutdownTimeout = 2 * time.Second` this is impossible.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go (hunk 20 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go]: renamed

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go (hunk 21 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+// Any new request(s) arriving after shutdownDelayDurationElapsedFn is closed
+// are replied with a 429 and the following response headers:
+//   - 'Retry-After: N` (so client can retry after N seconds, hopefully on a new apiserver instance)
+//   - 'Connection: close': tear down the TCP connection
+//
+// TODO: is there a way to merge WithWaitGroup and this filter?
+// NOTE: both WithRetryAfter and WithWaitGroup must use the same exact isRequestExemptFunc 'isRequestExemptFromRetryAfter,
+// otherwise SafeWaitGroup might wait indefinitely and will prevent the server from shutting down gracefully.
+func WithRetryAfter(handler http.Handler, shutdownDelayDurationElapsedCh <-chan struct{}) http.Handler {
+	return withRetryAfter(handler, isRequestExemptFromRetryAfter, newRetryAfterFunc(shutdownDelayDurationElapsedCh))
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:68]: nit: move the isRequestExemptFromRetryAfter comment above this line, out of the func doc

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go (hunk 8 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:67]: the `shutdownDelayDurationElapsedCh` is a detail of the caller, not the func. Would call the func `WithRetryAfterDuringShutdown(handler, shutdownCh)`. Then it is understandable without context.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func withRetryAfter(handler http.Handler, isRequestExemptFn isRequestExemptFunc, shouldRespondWithRetryAfterFn shouldRespondWithRetryAfterFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryAfterParams, sendRetryAfter := shouldRespondWithRetryAfterFn()
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:73]: `params, send := ...`. The func is about retryAfter. No need for fully qualified var names.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+		// Return a 429 status asking the client to try again after 5 seconds
+		w.Header().Set("Retry-After", "5")
+		http.Error(w, retryAfterParams.Message, http.StatusTooManyRequests)
+	})
+}
+
+// newRetryAfterFunc builds a shouldRespondWithRetryAfterFunc that starts
+// sending Retry-After response to all incoming requests once the shut down
+// delay duration (ShutdownDelayDuration) has elapsed.
+func newRetryAfterFunc(shutdownDelayDurationElapsedCh <-chan struct{}) shouldRespondWithRetryAfterFunc {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:98]: haven't checked where this is used outside of WithRetryAfter. Just inline this there. This abstraction does not help, but rather distributes complexity.

File: staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+		}
+	}
+}
+
+// isRequestExemptFromRetryAfter returns true if the given request should be exempt
+// from being rejected with a 'Retry-After' response.
+// NOTE: both 'WithRetryAfter' and 'WithWaitGroup' filters should use this function
+// to exempt the set of requests from being rejected or tracked.
+func isRequestExemptFromRetryAfter(r *http.Request) bool {
+	if isKubeApiserverUserAgent(r) || hasExemptPathPrefix(r) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go:119]: return isKubeApiserverUserAgent(r) || hasExemptPathPrefix(r)
